### PR TITLE
[WIP] Named window getter

### PR DESCRIPTION
**Pull request by jdm**
This is a rebase of https://github.com/servo/servo/pull/19904 with an attempt at porting Gecko's named property getter support to Servo on top of it. It's very rough but I want to see what effect it has on tests so far.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/21869)

### CODE REVIEW

**Comment by jdm** [comment]
I've reviewed all of the previous failures and made a couple code correctness changes, filed https://github.com/web-platform-tests/wpt/pull/13373 and https://github.com/web-platform-tests/wpt/pull/13372 for tests that are broken, and we now get many more realistic test results for many tests. I'm also going to finish the document/window getters related to iframes which should yield a bunch of additional improvements.

**Comment by jdm** [comment]
One interesting failure that I just noticed:
```
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/_webgl/conformance/glsl/misc/shader-uniform-packing-restrictions.html", "line": 209785, "action": "test_result", "expected": "OK"}
```
That's because the test has `<div id="description"></div>` and then inside the script block has `description();`, I presume.

**Comment by jdm** [comment]
I lied, that is not why the test times out. I wonder if it's just doing a lot of bare window lookups and the introduction of the named properties method is inhibiting the jit? I have a todo about one thing that probably affects that, so it's worth checking.

File: components/script/dom/window.rs [hunk 1]
```diff
@@ -1208,6 +1213,67 @@ impl WindowMethods for Window {
...
+            if named_elements.next().is_none() {
+                return Some(NonNull::new_unchecked(
+                    first_element.reflector().get_jsobject().get()
+                ));
+            }
+        }
+
+        let collection =
+            HTMLCollection::create(self, root, Box::new(filter));
+        Some(NonNull::new_unchecked(collection.reflector().get_jsobject().get()))
```
Comment (nox) [inline components/script/dom/window.rs]: You can just use `NonNull::new` here.

**Comment by pshaughn** [comment]
This looks like it's been dormant for a while, might it be something I could help with?

**Comment by jdm** [comment]
Yes! This is definitely a high value web compatibility feature. I did a rebase of this branch recently; I don't have access to the machine with the branch, unfortunately, but it wasn't too difficult, iirc.

**Comment by pshaughn** [comment]
@jdm Is that machine forever gone or can you get to it sometime this week? The rebase is harder for me than it was for you, since I don't know what the source code in 2018 was doing and some things seem to have moved between js::jsapi and js::rust in confusing ways since then.

**Comment by jdm** [comment]
I'll try and repeat the rebase.

**Comment by jdm** [comment]
I have rebased the branch such that it now compiles successfully.

**Comment by jdm** [comment]
https://github.com/jdm/rust-mozjs/tree/proxyopt2 is a rust-mozjs branch that contains a commit that is necessary in 0994d2242e8ecb65fc3abdebee04eaf7cbaa9ff7.

**Comment by pshaughn** [comment]
Is it possible you mean a different branch? I'm not finding a version of NewProxyObject that takes a boolean as its last argument.

**Comment by jdm** [comment]
I had forgotten to commit my changes. I've now pushed the latest commit.

**Comment by pshaughn** [comment]
Implementing Document.SupportedPropertyNames (#7273), including a hashmap of names, seems like a good next move here. Without that hashmap, the idiomatic case where a script refers to the same element over and over by its id will force a new tree walk on every reference; this tree walk might be the cause of the timeouts mentioned earlier.

**Comment by pshaughn** [comment]
What's the purpose of adding the singleton option in the proxyopt2 branch? Does it have a semantic effect or is it about optimization?

**Comment by pshaughn** [comment]
@jdm I'd like a little help, if you can remember what you were going for in window_named_properties.rs:
* how do I use the bool return code and *mut ObjectOpResult to make ProxyTraps do what it has to do on the methods that have those? In particular, what combination of them returns a Javascript-false without throwing an error?
* How do I build the AutoIdVector own_property_keys needs, once I have the list of Rust strings to put into it? 
* What's the relevance of the comments about a "class pointer"? Can we read https://github.com/ricardoquesada/Spidermonkey/blob/4a75ea2543408bd1b2c515aa95901523eeef7858/js/src/jsproxy.h#L40-L42 as permission to just not have a class pointer, or is the purpose of the class pointer something else?

**Comment by pshaughn** [comment]
I figured out the AutoIdVector part, since codegen is doing the exact same thing to hook up other SupportedPropertyNames cases already. The other methods only show up if a script is directly interacting with the window named properties object, which is certainly possible but is weird and might not even be in WPT.

I have another question now: How do I deal with the presence of this rust-mozjs patch with respect to getting things into taskcluster for testing? I'm at a point now where I'd like to see full test-battery results on what I've done, and my local machine both takes a long time getting them and times out in some tests that taskcluster doesn't.

**Comment by pshaughn** [comment]
^ that's my progress so far if you want to look; commit history-wise it's based on my doc named getter with the window getter code just pasted in manually, since there were enough conflicts that doing it all by hand was easier than trying to merge and then conflict-resolve

**Comment by jdm** [comment]
You can push a change to Cargo.toml that uses a git override like this for mozjs:
```
mozjs = { git = "https://github.com/jdm/servo", branch = "proxyopts2" }
```
and taskcluster will happily use it.

**Comment by jdm** [comment]
As for the other questions:
* we should add a rust method that replicates the effects of https://github.com/servo/mozjs/blob/b2f83932fe9d361face14efd03f2465b9262e687/mozjs/js/public/Class.h#L153-L170 in [this code](https://github.com/servo/mozjs/blob/e21c05b415dfc246175ff8d5fc48b0e8c5b4e9e9/mozjs/js/rust/src/lib.rs#L44)
* various callers of the helpers in https://searchfox.org/mozilla-central/rev/e076e40ab1290f4e5e67ebd21dc8af753fc05be6/js/src/jsapi.cpp#246 in Firefox suggest that the right way to use this is to return true and set the error code appropriately (returning false is what causes an exception to occur)
* the class is expected to be non-null by the engine, and it looks like it's mostly just coincidence that we don't crash right now given https://searchfox.org/mozilla-central/source/js/src/vm/ProxyObject.cpp#54-80 (ie. a debug build would dereference a null pointer; a non-singleton proxy would dereference a null pointer; etc.). We're going to need to copy what [Gecko does](https://searchfox.org/mozilla-central/rev/e076e40ab1290f4e5e67ebd21dc8af753fc05be6/dom/base/WindowNamedPropertiesHandler.cpp#212-223) and [pass the class](https://searchfox.org/mozilla-central/rev/e076e40ab1290f4e5e67ebd21dc8af753fc05be6/dom/base/WindowNamedPropertiesHandler.cpp#233) appropriately.

**Comment by mrobinson** [comment]
I'm going to close this since @delan has landed a later version of this change.

**Pull request closed without merging**